src/mozilla_sec_eia/extract.py

```diff
@@ -0,0 +1,244 @@
+
+logger = logging.getLogger(f"catalystcoop.{__name__}")
+
+
+class ExtractionMetadataSchema(pa.DataFrameModel):
+    """Define the required schema for extraction metadata.
+
+    Extra columns are permitted, but these are required for computing extraction metrics.
+    """
+
+    filename: pa.typing.Index[str] = pa.Field(check_name=True)
+    success: bool = pa.Field(coerce=True)
+
+
+def _load_artifact_as_csv(run: Run, artifact_name: str) -> pd.DataFrame:
+    """Download a CSV and parse to DataFrame from mlflow tracking server."""
+    df = pd.read_csv(
+        io.StringIO(mlflow.artifacts.load_text(run.info.artifact_uri + artifact_name))
+    )
+    return df
+
+
+def _log_artifact_as_csv(
+    artifact: pd.DataFrame, artifact_name: str, index: bool = True
+):
+    """Upload a DataFrame as a CSV to mlflow tracking server."""
+    return mlflow.log_text(artifact.to_csv(index=index), artifact_name)
+
+
+def _get_most_recent_run(experiment_name: str):
+    """Search mlflow for most recent extraction run with specified experiment name."""
+    run_metadata = mlflow.search_runs(experiment_names=[experiment_name])
+    return mlflow.get_run(
+        run_metadata[run_metadata["end_time"] == run_metadata["end_time"].max()][
+            "run_id"
+        ][0]
+    )
+
+
+def _get_filings_to_extract(
+    experiment_name: str,
+    metadata: pd.DataFrame,
+    continue_run: bool = False,
+    num_filings: int = -1,
+):
+    """Get filings that should be extracted by run.
+
+    Args:
+        experiment_name: Name of mlflow experiment.
+        metadata: Metadata for full set of filings to potentially extract.
+        continue_run: Whether to continue a previous extraction run.
+        num_filings: Number of filings to extract.
+    """
+    extraction_metadata = pd.DataFrame(
+        {"filename": pd.Series(dtype=str), "success": pd.Series(dtype=bool)}
+    ).set_index("filename")
+    extracted = pd.DataFrame()
+    run_id = None
+    if continue_run:
+        most_recent_run = _get_most_recent_run(experiment_name)
+        extraction_metadata = ExtractionMetadataSchema.validate(
+            _load_artifact_as_csv(
+                most_recent_run, "/extraction_metadata.csv"
+            ).set_index("filename")
+        )
+        extracted = _load_artifact_as_csv(most_recent_run, "/extracted.csv")
+        run_id = most_recent_run.info.run_id
+
+    filings_to_extract = metadata[~metadata["filename"].isin(extraction_metadata.index)]
+    if num_filings > 0:
+        filings_to_extract = filings_to_extract.sample(num_filings)
+    return (
+        filings_to_extract,
+        extraction_metadata,
+        extracted,
+        run_id,
+    )
+
+
+def _get_experiment_name(dataset: str, experiment_suffix: str | None = None) -> str:
+    experiment_name = f"{dataset}_extraction"
+    if experiment_suffix is not None:
+        experiment_name += f"_{experiment_suffix}"
+    return experiment_name
+
+
+def compute_validation_metrics(
```

Comment:
I could see this being a general function that orchestrates choosing and running a couple different smaller functions that actually calculate metrics. For example, for the Ex. 21 validation I think I want to have the following metrics for each unique filing ID:
* assert validation dataframe is the same as the calculated dataframe
* check the precision and recall for each column

But for the basic 10k extraction since we don't care about structure of the table that much, the assert dataframes equal check probably won't reveal much. 

I think I can implement some more complex/custom validations for the Ex. 21, but maybe there should be a structure in place in this PR to be able to add  validations functions and customize what gets run for each dataset. Does that make sense?

Reply:
Yeah that makes a lot of sense. I think it might be best for a separate PR though, so we can get this one over the finish line. I think a good sprint goal for me would be to work on an infrastructure focused PR to prep for the MVP, so maybe part of that could make the validation stuff more configurable.

Does that make sense to you?